Streamline campaign setup: Add confirmation modal when user skips without creating campaign during onboarding

asvinb opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes #2504 

_Replace this with a good description of your changes & reasoning._


### Screenshots:
<img width="613" alt="image" src="https://github.com/user-attachments/assets/d94f7b6d-e403-4453-a8ca-4f3a167475cd">




### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Go through the onboarding flow.
2. On the fourth step, click on the "Skip this step for now" button.
3. There should be a modal asking the user to confirm if they want to skip.
4. If the user clicks "Yes", the onboarding flow continues and the user is redirected to the dashboard.
5. If the user clicks "No", the modal is no longer displayed and the user stays on the page.


### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Add - Confirmation modal when user skips without creating a campaign during onboarding.

asvinb committed:
Add SkipPaidAdsConfirmationModal component.

asvinb committed:
Use correct function to complete onboarding.

asvinb committed:
Rename functions.

asvinb committed:
Update e2e tests.

asvinb commented on js/src/setup-mc/setup-stepper/setup-paid-ads/setup-paid-ads.js:
```diff
@@ -150,7 +161,13 @@ export default function SetupPaidAds() {
+						? finishOnboardingSetup
+						: handleShowSkipPaidAdsConfirmationModal
+				}
+				// TODO: might want to review eventName and eventProps
```
@joemcgill Shall we keep those props as is?

joemcgill replied:
Good question. I think we'll want to move that eventName to the new "Complete setup" button in the AYS modal, but the event name for this will likely need an update. Leave the TODO: comment for now and we can address that before the feature is merged.

asvinb replied:
👍

asvinb commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,93 @@
+ * Triggered when the skip button is clicked
+ * // TODO: to review
+ *
+ * @event gla_skip_paid_ads_modal_confirm_button_click
```
@joemcgill The event name TBC.

asvinb commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,93 @@
+				<AppButton
+					key="yes"
+					// TODO: confirm the eventName
+					eventName="gla_skip_paid_ads_modal_confirm_button_click"
```
@joemcgill If you can confirm please :)

asvinb commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,93 @@
+						Link: (
+							<AppDocumentationLink
+								href="https://support.google.com/google-ads/answer/10724817"
+								// TODO: review context and linkId values
```
@joemcgill Same here.

joemcgill commented on js/src/setup-mc/setup-stepper/setup-paid-ads/setup-paid-ads.js:
```diff
@@ -150,7 +161,13 @@ export default function SetupPaidAds() {
-				onClick={ finishOnboardingSetup }
+				onClick={
+					// Show the skip modal only for the "Skip this step for now" button only.
+					showPaidAdsSetup
+						? finishOnboardingSetup
+						: handleShowSkipPaidAdsConfirmationModal
```
Oh, it looks like there was confusion here. We're removing the initial "Skip this step for now" button from the `PaidAdsFeaturesSection` in #2500. This AYS modal should apply to any skip button (although there will only be the "Skip paid ads creation" button once that PR is merged).

asvinb replied:
👍  I've added TODOs to review the code once #2500 has been merged.

joemcgill commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,93 @@
+			title={ __( 'Skip setting up ads?', 'google-listings-and-ads' ) }
+			buttons={ [
+				<AppButton key="no" isSecondary onClick={ onRequestClose }>
+					{ __( 'No', 'google-listings-and-ads' ) }
```
Reference the Modal Content in the issue description, these buttons should "Cancel" and "Complete setup". I've updated the A/C to make that more clear.

asvinb replied:
👍

joemcgill reviewed (changes requested):
The functionality is working correctly when the "Skip..." button you've targeted is clicked, but this needs to be updated so this applies to all "Skip" buttons. Left feedback inline.

asvinb committed:
Address code review feedback.

joemcgill reviewed (approved):
This looks like it's working well. I'm not 100% confident in the way events need to be recorded, so we can confirm with @eason9487 and either update this PR after his review or spin up a follow-up issue to update all event tracking code for #2459.

ankitguptaindia commented:
Suggested some QA feedback on the connected issue here: https://github.com/woocommerce/google-listings-and-ads/issues/2504#issuecomment-2321093373

**Next Step:**
Moving PR to in-progress for feedback.

joemcgill commented:
Thanks @ankitguptaindia I've left a response to your feedback in https://github.com/woocommerce/google-listings-and-ads/issues/2504#issuecomment-2321601555. @asvinb Can you make the suggested text change to the modal button?

asvinb committed:
Update label.

asvinb commented:
Sure thing @joemcgill . Label updated @ankitguptaindia . Can you kindly check?

ankitguptaindia commented:
**QA/Test Report-** ✅ 
------------

**Testing Environment -** 

<Details>

* WordPress: 6.6.1
* Theme active on store: Twenty Twenty-Four Version: 1.2
* WooCommerce - Version 9.2.3 
* PHP: 8.3
* Web Server: Nginx
* Browser: Chrome - Version 127
* OS: macOS Sonoma 14.6.1 

</Details>





**Test Results -** The changes have been implemented as described. A confirmation modal will appear when the user clicks on ski ads. If the user clicks 'Cancel,' they remain on the same page. If the user clicks 'Complete setup without setting up ads,' the onboarding flow continues. 


**Functional Demo /  Screencast -** 


https://github.com/user-attachments/assets/aa5ba997-af15-4f5c-9f3e-897824e7e686

eason9487 commented on js/src/setup-mc/setup-stepper/setup-paid-ads/setup-paid-ads.js:
```diff
@@ -150,7 +161,9 @@ export default function SetupPaidAds() {
-				onClick={ finishOnboardingSetup }
+				onClick={ handleShowSkipPaidAdsConfirmationModal }
+				// TODO: Review eventName and eventProps
+				// The same eventName and eventProps has been copied over to the "Confirm" button in the SkipPaidAdsConfirmationModal component.
```
The event tracking related code can be removed from this button.

asvinb replied:
👍

eason9487 commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,125 @@
+	onSkipConfirmation = noop,
+	isProcessing = false,
+	showPaidAdsSetup = false,
+	paidAds = {},
```
If these props are expected to be specified, the default values may not be needed.

asvinb replied:
Default values removed @eason9487

eason9487 commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,125 @@
+				<AppButton
+					key="complete-setup"
+					// TODO: confirm the eventName
+					eventName="gla_onboarding_complete_button_click"
```
Suggest adding an extra click handler to `onSkipConfirmation` in `SetupPaidAds` component to send this event tracking with `eventProps` via `recordGlaEvent` util instead of sending it here. In this way, this modal doesn't need to be aware the states and data associated with the tracking.

asvinb replied:
Sounds good @eason9487 . I've added the tracking in `SetupPaidAds`
https://github.com/woocommerce/google-listings-and-ads/pull/2563/commits/afc0554a4892d38f5dd0dab9c49674efec7fb512#diff-944cb8c282b1af583fa6ba41e5509c7057f0f5c7c8838030b0ef0a33d8867d5cR146

eason9487 commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,125 @@
+ */
+const SkipPaidAdsConfirmationModal = ( {
+	onRequestClose = noop,
+	onSkipConfirmation = noop,
```
Suggest renaming this callback because the *confirmation* itself is not skipped.

asvinb replied:
Callback renamed to `onSkipCreatePaidAds`.

eason9487 commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,125 @@
+
+	return (
+		<AppModal
+			className="gla-ads-skip-paid-ads-modal"
```
The `gla-ads-skip-paid-ads-modal` class name seems unnecessary?

asvinb replied:
Removed @eason9487

eason9487 commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,125 @@
+				) }
+			</p>
+			<p>
+				{ createInterpolateElement(
```
It should not be necessary to use `createInterpolateElement` because the `Link` is not interpolated between the text.

asvinb replied:
👍 . Updated!

eason9487 commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,125 @@
+						Link: (
+							<AppDocumentationLink
+								href="https://support.google.com/google-ads/answer/10724817"
+								// TODO: review context and linkId values
+								context="skip-paid-ads-modal"
+								linkId="skip-paid-ads-modal-learn-more-performance-max"
```
```suggestion
								context="skip-paid-ads-modal"
								linkId="paid-ads-with-performance-max-campaigns-learn-more"
```

asvinb replied:
👍

eason9487 commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,125 @@
+ *
+ * @event gla_onboarding_complete_button_click
+ */
+
+/**
```
Adding tracking doc.

```suggestion

/**
 * @fires gla_documentation_link_click with `{ context: 'skip-paid-ads-modal', link_id: 'paid-ads-with-performance-max-campaigns-learn-more', href: 'https://support.google.com/google-ads/answer/10724817' }`
 */
 
/**
```

asvinb replied:
👍

eason9487 commented on js/src/setup-mc/setup-stepper/setup-paid-ads/skip-paid-ads-confirmation-modal.js:
```diff
@@ -0,0 +1,125 @@
+const SkipPaidAdsConfirmationModal = ( {
+	onRequestClose = noop,
+	onSkipConfirmation = noop,
+	isProcessing = false,
```
If the UI requirement doesn't ask for continuously showing this modal after confirming the skip, it's recommended to close the modal immediately instead of dealing with this state within the modal.

asvinb replied:
Updated @eason9487 . The modal is now closed when confirming the skip.

eason9487 reviewed (changes requested):
Thanks for the work.

Left some suggestions about reducing the complexity. Besides, the e2e testing couldn't pass - https://github.com/woocommerce/google-listings-and-ads/actions/runs/10697346716/job/29654336938

asvinb committed:
Address CR feedback.

asvinb committed:
Remove comment.
:

asvinb committed:
Fix E2E tests.

asvinb commented:
Thanks for the review @eason9487 . I fixed the failing E2E test which was after we changed the button label in the modal. Should be ok now:
https://github.com/woocommerce/google-listings-and-ads/actions/runs/10704832892

eason9487 commented on tests/e2e/utils/pages/setup-mc/step-4-complete-campaign.js:
```diff
@@ -179,4 +179,72 @@ export default class CompleteCampaign extends MockRequests {
+	async clickCancelModalButton() {
+		const button = this.getCancelModalButton();
+		await button.click();
+		await this.page.waitForLoadState( LOAD_STATE.DOM_CONTENT_LOADED );
```
Looks like this waiting is not necessary.

asvinb replied:
👍

eason9487 commented on tests/e2e/utils/pages/setup-mc/step-4-complete-campaign.js:
```diff
@@ -179,4 +179,72 @@ export default class CompleteCampaign extends MockRequests {
+	/**
+	 * Clicks the "Cancel" button in the skip paid ads creation modal.
+	 *
+	 * @return {Promise<void>} Resolves when the click action is completed and the page has loaded.
```
Does this action need to wait for any page loads?

asvinb replied:
Nope. Updated @eason9487

eason9487 commented on js/src/setup-mc/setup-stepper/setup-paid-ads/setup-paid-ads.js:
This line is no longer necessary.

https://github.com/woocommerce/google-listings-and-ads/blob/b884b97cc593f28900450637b01efffe56a64a7a/js/src/setup-mc/setup-stepper/setup-paid-ads/setup-paid-ads.js#L170

asvinb replied:
👍

eason9487 reviewed (approved):
Thanks for reducing the complexity. LGTM.

Left a few minor things to be tweaked and I believe it should not need another round of code reviews after adjusting them.

asvinb committed:
Fix unecessary page wait.

asvinb commented:
@eason9487 PR has been updated and ready for your final 👍

eason9487 reviewed (approved):
LGTM
